### ISSUE TITLE
load testing tags api

### ISSUE THREAD

**Issue by davecromberge** [Netflix/atlas#628]
I have been running load tests roughly based on [these Gatling tests](https://github.com/brharrington/atlas-load-test).

I have a mixed load simulation in place where a mixed load is generated as follows:

- 1,000,000 unique metrics are published at a 30second step interval.  The number of unique dimension keys for these metrics is < 20, in other words they have a relatively low cardinality.
- 10 users query the`/api/v1/tags` endpoint to list dimension keys (server limit is 1000), at a random interval between 10s and 120s.

The machine I am running my atlas server on is an r3.2x large, and have 2 m3.xlarge instances generating the publish and query load respectively.

My initial results show that most of the publish requests succeed.  However, the tags keys listing consistently 503s (the server request-timeout is set to 10s).  From the JMX metrics, I see queues building up in the `LocalDatabaseActor` instance - could this be the reason for the latency that I am observing? 

I have seen a different approach to fulfilling requests in the graph api, where an `ImperativeRequestContext` is used / actor per request model.  In the tags api, however, the ask pattern is used, with a timeout of 60 seconds.  If the queueing is indeed the reason for the latency, could it be because the default timeout of 60 seconds does not match the request timeout, causing the db actor mailbox to drain more slowly?

I am quite new to Atlas and am currently evaluating it - I would like to know if this is a reasonable scenario and if there is anything I can do to improve the results that I am seeing for listing metadata when the database is under heavy load from ingres.

**Comment by brharrington** [comment]
I'm just getting back to the office and need to catch up on some things so it will probably be several days before I can look at this closely. Quickly skimming through your description it sounds reasonable. Performance can vary quite a bit depending on data characteristics. If you can publish your tests I can try to reproduce.

**Comment by davecromberge** [comment]
Thank you @brharrington, I would really appreciate your input - I am quite happy to wait until you have some free time to look into it.

I have published my load tests here:
https://github.com/davecromberge/atlas-load-test/tree/mixed-load

The idea with the repository is to add further simulations for wide cardinalities, churning metrics etc in order to determine safe operating limits for various load conditions.  Therefore I have inlined the test input parameters into each simulation, while the endpoint config applies globally and is in the application.conf.

Furthermore, reports of my previous runs can be found [here](https://github.com/davecromberge/atlas-load-test/tree/mixed-load/src/test/scala/com/outlyer/atlas/loadtest/mixed/reports).  In several of these runs, the cardinalities were very high and I worked my way down from here.

I am running Atlas with the following [memory.conf](https://github.com/davecromberge/atlas-load-test/blob/mixed-load/src/test/resources/memory.conf) and JVM settings:
```
  -XX:+PreserveFramePointer
  -Xss1024k
  -XX:MetaspaceSize=256m
  -XX:MaxMetaspaceSize=256m
  -XX:+UseG1GC
  -XX:StringTableSize=1000003
  -Xms52g
  -Xmx52g
  -XX:+HeapDumpOnOutOfMemoryError
  -verbosegc
  -verbose:sizes
  -XX:+DisableExplicitGC
  -XX:+PrintGCDetails
  -XX:+PrintGCDateStamps
  -Xloggc:/usr/share/iep-atlas/logs/gc.log
```

Finally, the version of Atlas that I am using is 1.6-SNAPSHOT configured to use Spectator to publish internal metrics to JMX.

I have been using YourKit profiler to try find potential bottlenecks in [this code](https://github.com/Netflix/atlas/blob/master/atlas-core/src/main/scala/com/netflix/atlas/core/index/RoaringTagIndex.scala#L356-L365), but I suspect that latency could be a result of mailbox queues in the db actor, potentially due to misalignment of ask timeouts and request-timeouts as mentioned above.  If ask requests drained slower than requests arrive, this may lead to a backlog of messages to the LocalDatabaseActor.

Thank you again for your time.
